v2/lints/cabf_br/lint_san_dns_name_onion_invalid.go

```diff
@@ -0,0 +1,140 @@
+v2 addresses. As the EV Guidelines have, since v1.5.3, required that the CA
+obtain a demonstration of control from the Applicant, which effectively
+requires the `.onion` name to be well-formed, even prior to RFC 7686.
+
+See also https://github.com/cabforum/documents/issues/191
+*******************************************************************/
+
+package cabf_br
+
+import (
+	"encoding/base32"
+	"fmt"
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type onionNotValid struct{}
+
+func (l *onionNotValid) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true if the certificate contains one or more subject
+// names ending in `.onion`.
+func (l *onionNotValid) CheckApplies(c *x509.Certificate) bool {
+	// TODO(sleevi): This should also be extended to support nameConstraints
+	// in the future.
+	return util.CertificateSubjInTLD(c, util.OnionTLD)
+}
+
+// Execute will lint the provided certificate. A lint.Error lint.LintResult will
+// be returned if:
+//
+//  1) The certificate contains a Tor Rendezvous Spec v2 address and is not an
+//     EV certificate (BRs: Appendix C).
+//  2) The certificate contains a `.onion` subject name/SAN that is neither a
+//     Rendezvous Spec v2 or v3 address.
+func (l *onionNotValid) Execute(c *x509.Certificate) *lint.LintResult {
+	for _, subj := range append(c.DNSNames, c.Subject.CommonName) {
+		if !strings.HasSuffix(subj, util.OnionTLD) {
+			continue
+		}
+		labels := strings.Split(subj, ".")
+		if len(labels) < 2 {
+			return &lint.LintResult{
+				Status: lint.Error,
+				Details: fmt.Sprintf("certificate contained a %s domain with too "+
+					"few labels: %q", util.OnionTLD, subj),
+			}
+		}
+		onionDomain := labels[len(labels)-2]
+		switch len(onionDomain) {
+		// Onion v2 address. These are only permitted for EV, per BRs Appendix C.
+		// Per 2.4 of Rendezvous v2:
+		// Valid onion addresses contain 16 characters in a-z2-7 plus ".onion"
+		// This is handled later, during the Base32 decode
+		case 16:
+			if !util.IsEV(c.PolicyIdentifiers) {
+				return &lint.LintResult{
+					Status: lint.Error,
+					Details: fmt.Sprintf("%q is a v2 address, but the certificate is not "+
+						"EV", subj),
+				}
+			}
+		// Onion v3 address. Permitted for all certificates by CA/Browser Forum
+		// Ballot SC27.
+		//
+		// Per 1.2 of Rendezvous v3:
+		// A hidden service's name is its long term master identity key.  This is
+		// encoded as a hostname by encoding the entire key in Base 32, including a
+		// version byte and a checksum, and then appending the string ".onion" at the
+		// end. The result is a 56-character domain name.
+		case 56:
+		default:
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: fmt.Sprintf("%q is not a v2 or v3 Tor address", subj),
+			}
+		}
+		_, err := base32.StdEncoding.DecodeString(strings.ToUpper(onionDomain))
```

Comment:
Can you help me understand why the `strings.ToUpper` is required here? The example addrs in the onion spec seem to be all lowercase (e.g. `pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion`). If this is to normalize what could be a case insensitive `onionDomain` from the cert would it be less confusing to use `strings.ToLower`?

Reply:
Good question! I should have commented this, as it was otherwise magic.

The Base-32 specification (or at least, StdEncoding) works on RFC 4648, which defines the characterspace for Base-32 using purely uppercase characters. Lowercase would be distinct (as they are in base-64)/malformed. I added a comment to this effect, to make it clearer why it's being uppercased; any mixed case or lower case would be invalid base32

Reply:
That makes sense, thank you